pymmcore_plus/mda/_engine.py

```diff
@@ -220,57 +38,9 @@ def _prep_hardware(self, event: MDAEvent, waitForSystem: bool = True):
         if event.exposure is not None:
             self._mmc.setExposure(event.exposure)
 
-        if waitForSystem:
-            self._mmc.waitForSystem()
-
-    def _finish_run(self, sequence: MDASequence):
-        """
-        To be called at the end of an acquisition.
-
-        Parameters
-        ----------
-        sequence : MDASequence
-            The sequence that was finished.
-        """
-        logger.info("MDA Finished: {}", sequence)
-        self._running = False
-        self._events.sequenceFinished.emit(sequence)
-
-    def run(self, sequence: MDASequence) -> None:
-        """
-        Run the multi-dimensional acquistion defined by `sequence`.
-
-        Most users should not use this directly as it will block further
-        execution. Instead use ``run_mda`` on CMMCorePlus which will run on
-        a thread.
-
-        Parameters
-        ----------
-        sequence : MDASequence
-            The sequence of events to run.
-        """
-        try:
-            self._prepare_to_run(sequence)
-
-            for event in sequence:
-                cancelled = self._wait_until_event(event, sequence)
-
-                # If cancelled break out of the loop
-                if cancelled:
-                    break
-
-                logger.info(event)
-                self._prep_hardware(event)
-
-                self._mmc.snapImage()
-                img = self._mmc.getImage()
+        self._mmc.waitForSystem()
 
-                self._events.frameReady.emit(img, event)
-        except Exception as e:  # noqa E722
-            # clean up so future MDAs can be run
-            self._canceled = False
-            self._running = False
-            with contextlib.suppress(Exception):
-                self._finish_run(sequence)
-            raise e
-        self._finish_run(sequence)
+    def exec_event(self, event: MDAEvent) -> Any:
+        # TODO: add non-aquisition event-specific logic here later
+        self._mmc.snapImage()
+        return self._mmc.getImage()
```

Comment:
Can we specify a subclassable `MDAEventData` object? Then we could have an `image` or maybe `datatype` attributes. That way we can still support images, but also allow for other data types like my use case of collecting spectral data

Reply:
I like this idea, but what would you think about making it Protocol based rather than class based?

For instance, our internal implementation would just check for certain attributes on the object
```
image = getattr(obj, 'image', None)
```

Someone could then use a `NamedTuple`, or a `dataclass`  ... or a custom object.

Not 100% sure here on the implementation, but I definitely like the idea of returning a container object rather than a numpy array

Reply:
I changed this to returning a named tuple, with an attribute of `image`.  That allows the `_runner` logic to check for an `image` attribute (and only emit it if it's there)... and we can expand on that protocol later in the future.

Sound good? 

If so, how would you feel about this PR going in soon?